File: architecture-skills/design-patterns/examples/mediator/mediator.py
"""Mediator pattern in Python.

A chat room routes messages between users so users never hold direct
references to each other. Duck typing keeps it light: anything with a
`receive()` method can join the room.
"""

from __future__ import annotations


class ChatRoom:
    """Mediator: knows every participant and owns the routing rules."""

    def __init__(self) -> None:
        self._users: dict[str, User] = {}

    def register(self, user: User) -> None:
        self._users[user.name] = user
        user.room = self

    def broadcast(self, sender: User, text: str) -> None:
        for user in self._users.values():
            if user is not sender:
                user.receive(sender.name, text)

    def direct(self, sender: User, recipient: str, text: str) -> None:
        user = self._users.get(recipient)
        if user is None:
            sender.receive("room", f"no such user: {recipient}")
        else:
            user.receive(sender.name, f"(dm) {text}")


class User:
    """Colleague: talks only to the mediator."""

    def __init__(self, name: str) -> None:
        self.name = name
        self.room: ChatRoom | None = None

    def say(self, text: str) -> None:
        assert self.room is not None, "user must join a room first"
        self.room.broadcast(self, text)

    def dm(self, recipient: str, text: str) -> None:
        assert self.room is not None, "user must join a room first"
        self.room.direct(self, recipient, text)

    def receive(self, sender: str, text: str) -> None:
        print(f"[{self.name}] {sender}: {text}")


if __name__ == "__main__":
    room = ChatRoom()
    ada, grace, alan = User("Ada"), User("Grace"), User("Alan")
    for user in (ada, grace, alan):
        room.register(user)

    ada.say("Hello, everyone!")
    grace.dm("Ada", "Hi Ada, got a minute?")
    alan.dm("Linus", "ping")

File: architecture-skills/design-patterns/examples/mediator/mediator.js
// Mediator pattern in JavaScript.
//
// A chat room routes messages between users. In Node or the browser the
// same decoupling is often built on an EventEmitter or EventTarget acting
// as the mediator; the explicit class makes the routing rules visible.

class ChatRoom {
  #users = new Map();

  register(user) {
    this.#users.set(user.name, user);
    user.room = this;
  }

  broadcast(sender, text) {
    for (const user of this.#users.values()) {
      if (user !== sender) {
        user.receive(sender.name, text);
      }
    }
  }

  direct(sender, recipient, text) {
    const user = this.#users.get(recipient);
    if (!user) {
      sender.receive("room", `no such user: ${recipient}`);
    } else {
      user.receive(sender.name, `(dm) ${text}`);
    }
  }
}

class User {
  constructor(name) {
    this.name = name;
    this.room = null;
  }

  say(text) {
    this.room.broadcast(this, text);
  }

  dm(recipient, text) {
    this.room.direct(this, recipient, text);
  }

  receive(sender, text) {
    console.log(`[${this.name}] ${sender}: ${text}`);
  }
}

const room = new ChatRoom();
const ada = new User("Ada");
const grace = new User("Grace");
const alan = new User("Alan");
for (const user of [ada, grace, alan]) {
  room.register(user);
}

ada.say("Hello, everyone!");
grace.dm("Ada", "Hi Ada, got a minute?");
alan.dm("Linus", "ping");
